core/cat/rabbit_hole.py

```diff
@@ -1,33 +1,114 @@
 import os
 import time
-import tempfile
+import json
 import mimetypes
 from typing import List, Union
+from urllib.request import urlopen
+from urllib.parse import urlparse
 
 from cat.log import log
 from starlette.datastructures import UploadFile
-from langchain.document_loaders import (
-    PDFMinerLoader,
-    UnstructuredURLLoader,
-    UnstructuredFileLoader,
-    UnstructuredMarkdownLoader,
-)
+from fastapi import HTTPException
 from langchain.docstore.document import Document
+from qdrant_client.http import models
+
+from langchain.document_loaders.parsers import PDFMinerParser
+from langchain.document_loaders.parsers.generic import MimeTypeBasedParser
+from langchain.document_loaders.parsers.txt import TextParser
+from langchain.document_loaders.blob_loaders.schema import Blob
+from langchain.document_loaders.parsers.html.bs4 import BS4HTMLParser
 
 
 class RabbitHole:
     """
 
     """
+
     def __init__(self, cat):
         self.cat = cat
 
+        self.file_handlers = {
+            "application/pdf": PDFMinerParser(),
+            "text/plain": TextParser(),
+            "text/markdown": TextParser(),
+            "text/html": BS4HTMLParser()
+        }
+
+    def ingest_memory(self, file: UploadFile):
+        """Upload memories to the declarative memory from a JSON file.
+
+        Parameters
+        ----------
+        file : UploadFile
+            File object sent via `rabbithole/memory` hook.
+
+        Notes
+        -----
+        This method allows to upload a JSON file containing vector and text memories directly to the declarative memory.
+        When doing this, please, make sure the embedder used to export the memories is the same as the one used
+        when uploading.
+        The method also performs a check on the dimensionality of the embeddings (i.e. length of each vector).
+
+        """
+
+        # Get file bytes
+        file_bytes = file.file.read()
+
+        # Load fyle byte in a dict
+        memories = json.loads(file_bytes.decode("utf-8"))
+
+        # Check the embedder used for the uploaded memories is the same the Cat is using now
+        upload_embedder = memories["embedder"]
+        cat_embedder = str(self.cat.embedder.__class__.__name__)
+
+        if upload_embedder != cat_embedder:
+            message = f'Embedder mismatch: file embedder {upload_embedder} is different from {self.cat.embedder}'
+            raise HTTPException(
+                status_code=422,
+                detail={
+                    "error": message
+                })
```

Comment:
here I had to raise an `HTTPException` from inside the `RabbitHole`. How could I keep the exception in the endpoint?